inst/include/dust/dust.hpp

```diff
@@ -639,6 +808,201 @@ class Dust {
+      std::vector<uint64_t> rng(np * rng_len); //  Deinterleaved RNG state
+      // D -> H copies
+      _device_data.y.get_array(y);
+      _device_data.rng.get_array(rngi);
+#ifdef _OPENMP
+      #pragma omp parallel for schedule(static) num_threads(_n_threads)
+#endif
+      for (size_t i = 0; i < np; ++i) {
+        destride_copy(y_tmp.data(), y, i, np);
+        _particles[i].set_state(y_tmp.begin());
+
+        // Destride RNG
+        for (size_t j = 0; j < rng_len; ++j) {
+          rng[i * rng_len + j] = rngi[i + j * np];
+        }
+      }
+      _rng.import_state(rng);
+      _stale_host = false;
+    }
+  }
 };
 
+// TODO: The exact type here for the shared memory will likely change;
+// we'll need to do some translation in dust into native types in
+// order to make the device copy possible. There's nothing really
+// complicated in these in practice.
+template <typename T>
+void run_particles(size_t step_start, size_t step_end, size_t n_particles,
+                   size_t n_state, size_t n_int, size_t n_real, size_t n_pars,
+                   typename T::real_t * state,
+                   std::vector<dust::shared_ptr<T>> shared,
+                   uint64_t * rng_state) {
+  typedef typename T::real_t real_t;
+  const size_t n_particles_each = n_particles / n_pars;
+
+  // Allocation for our internal storage. We won't ever move this to
+  // or from the device so we just pass the size.
+  std::vector<real_t> state_next_v(n_particles * n_state);
+  std::vector<int> internal_int_v(n_particles * n_int);
+  std::vector<real_t> internal_real_v(n_particles * n_real);
+  real_t * state_next = state_next_v.data();
+  int * internal_int = internal_int_v.data();
+  real_t * internal_real = internal_real_v.data();
+  real_t * ret = state; // for ease of bookkeeping on return
+
+  // omp here
+  for (size_t i = 0; i < n_particles; ++i) {
+    dust::interleaved<real_t> p_state(state, i, n_particles);
+    dust::interleaved<real_t> p_state_next(state_next, i, n_particles);
+    dust::interleaved<int> p_internal_int(internal_int, i, n_particles);
+    dust::interleaved<real_t> p_internal_real(internal_real, i, n_particles);
+    dust::interleaved<uint64_t> p_rng(rng_state, i, n_particles);
+    // TODO: this needs work before moving to the device, but it might
+    // not be that bad in practice.
+    dust::shared_ptr<T> p_shared = shared[i / n_particles_each];
+
+    dust::rng_state_t<real_t> rng_block = dust::get_rng_state<real_t>(p_rng);
+    for (size_t step = step_start; step < step_end; ++step) {
+      update_device<T>(step,
+                       p_state,
+                       p_internal_int,
+                       p_internal_real,
+                       p_shared,
+                       rng_block,
+                       p_state_next);
+      std::swap(p_state, p_state_next);
+      // dust::interleaved<real_t> tmp = p_state;
+      // p_state = p_state_next;
+      // p_state_next = tmp;
```

Comment:
There is a swap in thrust, but not CUDA, so we probably want to revert this change

Reply:
we could implement our own :)

Reply:
Leaving this as is for now as we can change that around easily later